### ISSUE TITLE
Overall Question - Walkthrough without Ab1 conversion

### ISSUE THREAD

**Issue by rleenay** [markrobinsonuzh/crispRvariants#4]
Hello,

I am attempting to utilize this program to analyze some amplicon data, but I'm having an issue parsing through this first few blocks of code because I don't need to convert anything from ab1 files. 

Do you by chance have another example script you could post or send me to compare to? Possibly something directly from a fastq file to bam to plotting?

Thank you!

**Comment by HLindsay** [comment]
Hi,

I need a few more details to help you with this.  Are you planning to map your reads to a standard genome or an amplicon sequence?  Do you have access to a computer with an aligner (e.g. bwa) already installed?  If not, what operating system are you using?  Are your sequenced reads single-end or paired-end?  If the mapping step is new to you it might be easiest to do it using Galaxy - https://galaxyproject.org/main/ and start from bam files in CrispRVariants.

Best wishes,
Helen

**Comment by rleenay** [comment]
Helen,

Thank you for the rapid reply! The issue with the mapping was just the syntax in R. I puzzled my way through most of it - my issue was that the file structure in R was confusing to me at first for me to re-work that to a working directory in R versus pulling from the CrispRVariants example. I think it would have been helpful to see a slightly different example with some local filenames, but I appreciate the help. I'll follow-up if I find any other problems!

**Comment by rleenay** [comment]
Hello again Helen,

As a quick follow-up, I've gotten the overall pipeline working but I'm having a significant issue with the "CollapsePairs" function - apparently R cannot locate the function. All other functions are calling successfully, including plotting individual .bam files from forward and reverse reads. However, I have yet to figure out how I can take two of my aligned .bam files and combine them into a single consensus prior to generating the "crispr_set" object

![screen shot 2017-10-17 at 1 57 21 pm](https://user-images.githubusercontent.com/32717391/31689065-217bdc24-b343-11e7-8b0b-7ea1241568aa.png)

**Comment by HLindsay** [comment]
Hi,

The `collapsePairs` function is an internal function called by `readsToTarget`.  To use it, set the `collapse.pairs` parameter in `readsToTarget`, e.g.

```R
crispr_set <- readsToTarget(bam.files, guide, reference = ref,
                            target.loc = 22, collapse.pairs = TRUE)
```
This checks that both members of a pair have the same insertions and deletions if they both overlap the guide region.  

Best wishes,
Helen

**Comment by rleenay** [comment]
Is there an advantage of running the collapse.pairs and mapping the reads separately versus running "bwa mem" prior to creating crispr_set?

**Comment by HLindsay** [comment]
I'm not sure if I've understood your question correctly, let me know if this doesn't answer it:

`readsToTarget` always works on a bam file (mapped reads).  With `collapse.pairs = TRUE`, it assumes that the reads have been mapped as a pair and that this is indicated in the the flag column of the bam file.  If you have paired reads and don't set `collapse.pairs = TRUE`, it will interpret every read as a duplication and they'll appear in the "Other" column of your plots.  If you mapped the fastq files separately they would be in separate bam files.  I'm not sure offhand what would happen if you merged these into a single file and then ran CrispRVariants, as the flag column would indicate they are single end reads.

 It's possible to merge the pairs prior to mapping, e.g. with PEAR https://sco.h-its.org/exelixis/web/software/pear/.  We tried this in the CrispRVariants paper and found that merging prior to mapping resulted in slightly more mapped reads but didn't change the major variant alleles we saw.

**Comment by rleenay** [comment]
That does clear things up - thank you!

I wasn't sure if collapse.pairs required that the two .bam files be from seperate .fastq maps to the same reference, which is why I was running into an issue. I'll change the bwa command to map them as paired reads which should fix my issue.

**Comment by rleenay** [comment]
One quick question - I'm having a bit of trouble with your orientation settings for plotVariants().

My reference strand appears to be showing up correctly, but unfortunately my reads are being displayed reverse-complement. All of the options (orientation=opposite, or changing the reference sequence to RevCom) seem to affect my reference strand but not the actual reads. Could you possibly shed some light on how the plotVariants/readsToTarget method determines the orientation of the reads/variants?

Thank you again! This is extremely helpful. 
 
![screen shot 2017-10-18 at 8 44 37 am](https://user-images.githubusercontent.com/32717391/31728276-b18707ba-b3e0-11e7-9523-46cab65c56ae.png)

**Comment by HLindsay** [comment]
Thanks for showing me this, this should raise an error!  I'll fix this before the new release at the end of this month.  

Here's a small example of the most common use case.  If you run this it will hopefully show you how to set things up for your data:

```R
library(GenomicAlignments)
library(Biostrings)
library(CrispRVariants)

ex1_file <- system.file("extdata", "ex1.bam", package="Rsamtools")
gal <- readGAlignments(ex1_file, param=ScanBamParam(what=c("flag","seq")),
                       use.names = TRUE)
gal <- GAlignmentsList("sample1" = gal)

target <- GRanges("seq1", IRanges(5,27), strand = "+")
negative_target <- GRanges("seq1", IRanges(5,27), strand = "-")

ref <- DNAString("AGTGGCTCATTGTAAATGTGTGG")
ref_rc <- Biostrings::reverseComplement(ref)

# Reference and guide are on the positive strand
c1 <- readsToTarget(gal, target = target, reference = ref, target.loc = 17)
plotAlignments(c1)

# Reference and guide are on the negative strand
c2 <- readsToTarget(gal, target = negative_target, reference = ref_rc, target.loc = 17)
plotAlignments(c2)
```
Typically, the reference sequence would always end in "NGG", or contain "NGG" if you're counting in a larger region.  If the reference sequence maps to the negative strand of the genome, then set `strand = "-"` in the target and the reads will be displayed with respect to the negative strand.  `orientation = "target"` (the default) means use the strand information of the target.  

The other `orientation` options are for cases where, for example, you have targeted a gene on the negative strand with a guide on the positive strand and you want to number the variants with respect to the gene.  The parameter `reverse.complement` was in the original release, which only allowed two options - as is or reverse complement.   `orientation` and `reverse.complement` are largely redundant, but I've kept `reverse.complement` for now to avoid breaking peoples' code.  

Best,
Helen

**Comment by rleenay** [comment]
Helen,

Thank you again for that walkthrough - it really helped me figure out how to get this running. 

In terms of the reverse complement issue, I actually figured out another workaround:

'for(i in 1:length(bm_fnames)) {
strand_test <- strand(gdl)[i]
if ("-" %in% strand_test) {
  temp_reference=reference[i]
  rev_comreference=Biostrings::reverseComplement(Biostrings::DNAString(temp_reference))
  reference[i]=toString(rev_comreference)
}
}'

Where I can iterate through the gdl file, pull the strandedness of the guide from the called .bed file, and convert that to reverse complement using the Biostrings function. This allowed me to create multiple reference sequences, all oriented with the NGG PAM on the "right" side of the plot

**Comment by HLindsay** [comment]
Hi,

Just a couple of quick things - you can simplify your code a bit here as these functions are vectorized, e.g. 

```R
library(GenomicRanges)
library(Biostrings)

guides <- GRanges(c("chr1", "chr2", "chr3"),
                  IRanges(start = c(5,10,15), width = 5),
                  strand = c("-","-","+"))      
references <- DNAStringSet(c("AAAAA","CCCCC","GGGGG"))
neg_strand <- strand(guides) == "-"
references[neg_strand] <- reverseComplement(references[neg_strand])
```
It's fine to input the reference as a `DNAString`.  In fact it's converted if it's input as a character string.  In the example, `references[[1]]` gives you the first sequence as a `DNAString`.

It doesn't look like this is the case, but if you have sequences from different guides in the same bam file, consider using `readsToTargets` instead of `readsToTarget`.

Best,
Helen

**Comment by rleenay** [comment]
I will definitely look into that to see if that simplifies my code. If I remember correctly, I had to convert the 'DNAString' back to a character string because it wouldn't let me insert the 'DNAString' into any subsequent referenece[i] position. That being said, it definitely could have been a syntax error, as I am new to R programming. I'll look into it.

In terms of different guides, that's not quite the case but I'll keep that in mind for future datasets!

Thank you again!

-RTL

**Comment by rleenay** [comment]
Helen,

I have one (final - hopefully) question. I'm attempting to generate multiple "crispr_set"'s at the same time, but unfortunately I get an error in my for loop:

`object of type 'S4' is not subsettable`

How can I iterate through multiple file-names to generate multiple crispr_sets?

Thank you!

**Comment by HLindsay** [comment]
Hi,

That error might mean you should be subsetting with single instead of double brackets. 

I usually start with a named `GRanges` object listing the guides, a named `DNAStringSet` with the references and a named list with the matching bam filenames.  Here's some example code, in practise I'd read the guide locations from a bed file and read the reference sequences from a fasta file or extract them from a genome:

```R
library(CrispRVariants)
library(Biostrings)
library(GenomicRanges)

guides <- GRanges("chr1", IRanges(c(1,10), width = 20))
names(guides) <- c("A", "B") 
# (if you import a bed file from rtracklayer you could use the "name" metadata column instead)

bam_filenames <- list("A" = c("a1.bam", "a2.bam"), "B" = c("b1.bam","b2.bam"))

references <- DNAStringSet(c("AAAAAAAAAA","CCCCCCCCCC"))
names(references) <- c("A", "B")

csets <- lapply(names(guides), function(nm){
                   guide <- guides[nm]  # Single bracket subscript for GRanges
                   reads <- bam_filenames[[nm]] # Double bracket for lists
                   ref <- references[[nm]]  # Double bracket to get a DNAString from a DNAStringSet
                   cset <- readsToTarget(reads, guide, reference = ref, target.location = 5)
})
```

**Comment by rleenay** [comment]
Helen, I see what you are doing with lapply, but is there anyway to do it within a nested `for()` loop? I've got my references, bam filenames, and then .bed file indices cross-referenced in a data-frame so I can easily keep track of all experimental replicates - but when I run the for loop I cannot iterate through a list of filenames because of the S4 object limitation of the `readstoTarget()` function. This is an example of the loop:

`crispr_set_filenames[i] <- readsToTarget(bam_fnames, target = gdl[(MergedMaster[i,2])], reference = MergedMaster[i,3], target.loc = 22, collapse.pairs=TRUE, reverse.complement=TRUE, orientation = "target")`

Where MergedMaster contains these cross-referenced files/sequences for the `readsToTarget()`function.

Thank you again!

**Comment by HLindsay** [comment]
Hi

I can't really tell what's going on here without knowing what `MergedMaster` is.  What is the output of these commands?

```R
i <- 1
gdl[(MergedMaster[i,2])]
MergedMaster[i,3]
```
Your code passes the same set of bam filenames for every value of `i`, is this what you intend?

By the way, the issue you pointed out with the plots when the target and reference are on opposite strands is now fixed in version 1.5.9.

**Comment by rleenay** [comment]
Yes, that was the intention. 

I had an issue writing the "crispr_set" variable to a vector of crispr_set filenames, but I was able to make it work through the creation of a "crisper_set_list" list variable. I think the problem came up because `ReadsToTarget()` outputs the S4 type variable which only was able to be written to a list. 

After reading more, it's possible I could have also used the `purrr` package to iterate across multiple variables.

To answer your question - the gdl[(MergedMaster[i,2])] outputs the standard GRanges output, the MergedMaster column 2 is an index calling the respective target site from the gdl file. The MergedMaster[i,3] is a column of reference sequences (in character format).
